src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs

```diff
@@ -738,19 +742,35 @@ override public object ExecuteScalar()
             // between entry into Execute* API and the thread obtaining the stateObject.
             _pendingCancel = false;
 
-            SqlStatistics statistics = null;
-
+            Guid operationId = _diagnosticListener.WriteCommandBefore(this, nameof(ExecuteScalar));
 
+            SqlStatistics statistics = null;
+            
+            Exception e = null;
             try
             {
                 statistics = SqlStatistics.StartTimer(Statistics);
                 SqlDataReader ds;
                 ds = RunExecuteReader(0, RunBehavior.ReturnImmediately, returnStream: true);
                 return CompleteExecuteScalar(ds, false);
             }
+            catch (Exception ex)
+            {
+                e = ex;
+                throw;
+            }
             finally
             {
                 SqlStatistics.StopTimer(statistics);
+
+                if (e != null)
+                {
+                    _diagnosticListener.WriteCommandError(operationId, this, nameof(ExecuteScalar), e);
+                }
+                else
+                {
+                    _diagnosticListener.WriteCommandAfter(operationId, this, nameof(ExecuteScalar));
```

Comment:
@stephentoub In my personal opinion, the clear delineation between `*After` and `*Error` in the API is a true portrayal of not only the different logged events, but also give us two different code paths to minimize spaghetti logic.  The data points collected for these two events only have a couple of things in common, whereas the `*Error` event collects [`Exception`](https://github.com/tstringer/corefx/blob/sqlclient-diagnostics/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs#L67) and the `*After` event collects [`Statistics`](https://github.com/tstringer/corefx/blob/sqlclient-diagnostics/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs#L50).  Whereas I can understand your thoughts to merge this into a single method, I think the entry point is really the only common part of these two events.

In summary, going forward I think having both methods gives a clear indication of future calls.  With that being said, I'd be curious to get @avanderhoorn's thoughts on this.